service/history/api/pollupdate/api_test.go

```diff
@@ -118,36 +135,88 @@ func TestPollOutcome(t *testing.T) {
-					RunId:      t.Name() + "-run-id",
+					WorkflowId: workflowId,
+					RunId:      runId,
 				},
 				UpdateId: updateID,
 			},
+			WaitPolicy: &updatepb.WaitPolicy{
+				LifecycleStage: enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_COMPLETED,
+			},
 		},
 	}
 
 	t.Run("update not found", func(t *testing.T) {
 		reg.FindFunc = func(ctx context.Context, updateID string) (*update.Update, bool) {
 			return nil, false
 		}
-		_, err := pollupdate.Invoke(context.TODO(), &req, wfcc)
+		_, err := pollupdate.Invoke(context.TODO(), &req, shardContext, wfcc)
 		var notfound *serviceerror.NotFound
 		require.ErrorAs(t, err, &notfound)
 	})
-	t.Run("future timeout", func(t *testing.T) {
+	t.Run("context deadline expiry before server-imposed deadline expiry", func(t *testing.T) {
 		reg.FindFunc = func(ctx context.Context, updateID string) (*update.Update, bool) {
 			return update.New(updateID), true
 		}
-		ctx, cncl := context.WithTimeout(context.Background(), 5*time.Millisecond)
+		ctx, cncl := context.WithTimeout(context.Background(), serverImposedTimeout/2)
 		defer cncl()
-		_, err := pollupdate.Invoke(ctx, &req, wfcc)
+		_, err := pollupdate.Invoke(ctx, &req, shardContext, wfcc)
 		require.Error(t, err)
 	})
+	t.Run("context deadline expiry after server-imposed deadline expiry", func(t *testing.T) {
+		reg.FindFunc = func(ctx context.Context, updateID string) (*update.Update, bool) {
+			return update.New(updateID), true
+		}
+		ctx, cncl := context.WithTimeout(context.Background(), serverImposedTimeout*2)
+		defer cncl()
+		resp, err := pollupdate.Invoke(ctx, &req, shardContext, wfcc)
+		require.NoError(t, err)
+		require.Nil(t, resp.GetResponse().Outcome)
+		require.Equal(t, enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED, resp.Response.GetStage())
+	})
+	t.Run("non-blocking poll with omitted/unspecified wait policy", func(t *testing.T) {
+		for _, req := range []historyservice.PollWorkflowExecutionUpdateRequest{{
+			Request: &workflowservice.PollWorkflowExecutionUpdateRequest{
+				UpdateRef: req.Request.UpdateRef,
+			},
+		}, {
+			Request: &workflowservice.PollWorkflowExecutionUpdateRequest{
+				UpdateRef: &updatepb.UpdateRef{
+					WorkflowExecution: &commonpb.WorkflowExecution{
+						WorkflowId: workflowId,
+						RunId:      runId,
+					},
+					UpdateId: updateID,
+				},
+				WaitPolicy: &updatepb.WaitPolicy{
+					LifecycleStage: enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_UNSPECIFIED,
+				},
+			},
+		}} {
+			reg.FindFunc = func(ctx context.Context, updateID string) (*update.Update, bool) {
+				return update.New(updateID), true
+			}
+			resp, err := pollupdate.Invoke(context.Background(), &req, shardContext, wfcc)
+			require.NoError(t, err)
+			require.True(t, len(resp.GetResponse().UpdateRef.GetWorkflowExecution().RunId) > 0)
+			require.Nil(t, resp.GetResponse().Outcome)
+			require.Equal(t, enumspb.UPDATE_WORKFLOW_EXECUTION_LIFECYCLE_STAGE_ADMITTED, resp.Response.GetStage())
```

Comment:
> I would test an incomplete async update to verify that this is indeed non-blocking.

I think that this test provides sufficient coverage of that: here the Update is at Admitted stage, and there is no possibilty that the Update state machine can advance. Therefore the fact that the History service `Invoke` handler (when given a request that does not specify `WaitPolicy`) returns, and returns the expected lifecycle stage, shows that it did not block until Accepted/Completed.

Reply:
Confirmed with @bergundy offline -- we agree that tests are covering the non-blocking nature.